package/setup.py

```diff
@@ -130,33 +129,93 @@ def get(self, option_name, default=None):
         except configparser.NoOptionError:
             return default
 
-class MDAExtension(Extension, object):
-    """Derived class to cleanly handle setup-time (numpy) dependencies.
+class MDADistribution(Distribution, object):
+    """Derived class to cleanly handle setup-time (cython/numpy) dependencies.
     """
-    # The only setup-time numpy dependency comes when setting up its
-    #  include dir.
-    # The actual numpy import and call can be delayed until after pip
-    #  has figured it must install numpy.
-    # This is accomplished by passing the get_numpy_include function
-    #  as one of the include_dirs. This derived Extension class takes
-    #  care of calling it when needed.
-    def __init__(self, *args, **kwargs):
-        self._mda_include_dirs = []
-        super(MDAExtension, self).__init__(*args, **kwargs)
-
-    @property
-    def include_dirs(self):
-        if not self._mda_include_dirs:
-            for item in self._mda_include_dir_args:
-                try:
-                    self._mda_include_dirs.append(item()) #The numpy callable
-                except TypeError:
-                    self._mda_include_dirs.append(item)
-        return self._mda_include_dirs
-
-    @include_dirs.setter
-    def include_dirs(self, val):
-        self._mda_include_dir_args = val
+    # Setup-time cython dependency comes when cythonizing sources (cython might
+    #  not be available) or configuring numpy's include dir (numpy might
+    #  not be available either).
+    # The actual cython/numpy imports and calls can be delayed until after
+    #  pip/setuptools have figured they must install them.
+    # This is accomplished by using a distutils.core.Distribution subclass
+    #  that lazily fills-in cython and numpy information by hooking into
+    #  the run_commands method.
+    def __init__(self, attrs=None):
+        # Attributes must exist if they're to be imported from attrs.
+        self.mda_use_cython = is_dev
+        self.mda_build_requires = None
+        super(MDADistribution, self).__init__(attrs)
+
+    def run_commands(self):
+        # The following seems fragile...
```

Comment:
If we push this to build-time (which would be the cleanest, IMO), then even `pip install MDAnalysis` will fail if you don't have  `biopython` and `numpy`. What happens is the following:
 1. `pip` probes for `MDAnalysis` dependencies using `egg_info`; the `setup.py` creates the `egg-info` dir. `pip` learns it must install `numpy`, `biopython` and others;
 2. Before installing anything `pip` must learn if the dependencies themselves have dependencies. Thus, `./setup.py egg_info` will be run for every dependency, including `biopython`;
 3. Not being as awesome as ours, the `setup.py` of `biopython` proceeds to bork, because it needs `numpy` even to list its dependencies.

We solve this by doing the `numpy` install ourselves as soon as we suspect `pip` is in use.
This, of course, is `biopython`'s fault, but if we can make life easier for our users, better.

It is still not foolproof, though: if there's a package `mdthingy` that depends on `mdanalysis` and `biopython` our hack will only save the install if `pip` probes us before `biopython`.